### PR TITLE
mordenize cloud build yaml files with script 

### PR DESCRIPTION
**Pull request by smeet07**
addresses #790

### CODE REVIEW

**Comment by smeet07** [comment]
@grayside

**Comment by pattishin** [comment]
Hi @smeet07, appreciate the PR! Do you mind checking out the current failure in `api-unit-tests`?

Screenshot 2023-06-15 at 3 10 53 PM

**Comment by pattishin** [comment]
Hi @smeet07, appreciate the change! It looks like the `api-unit-tests` did not pass. Do you mind looking into it?

**Comment by smeet07** [comment]
> Hi @smeet07, appreciate the change! It looks like the `api-unit-tests` did not pass. Do you mind looking into it?

It says you need permissions to view the details

**Comment by smeet07** [comment]
@pattishin

**Comment by rogerthatdev** [comment]
@smeet07 
Thanks for the contribution!

In my testing I found that there is an issue having to do with variable substitution in your changes. For example, when initiating the `web-new-build` Cloud Build trigger that uses config `ops/web-build.cloudbuild.yaml` it returns the following error:
```
invalid argument "-docker.pkg.dev//website/website:" for "-t, --tag" flag: invalid reference format
See 'docker build --help'.
ERROR
ERROR: build step 0 "gcr.io/cloud-builders/docker" failed: step exited with non-zero status: 125
```

`-docker.pkg.dev//website/website:` in the build config reads `${_REGION}-docker.pkg.dev/${PROJECT_ID}/website/website:${_IMAGE_TAG} ` but the `script` field does not  support values from the [`substitution` field of the build config](https://github.com/GoogleCloudPlatform/emblem/blob/4be1a2f76f7b684f427d74db0209506e3d60f5af/ops/web-build.cloudbuild.yaml#L30C1-L32C25) 

Instead, each step that uses the `script` field with substitution values needs to also include an `env` field that will pass values to the script via environment variables. For example, for the `ops/web-build.cloudbuild.yaml`: 
```
steps:
  # Docker Build 
  # Copying the client-libs for Docker context
  - name: gcr.io/cloud-builders/docker
   env:
    - '_REGION=$_REGION'
    - 'PROJECT_ID=$PROJECT_ID'
    - '_IMAGE_TAG=$_IMAGE_TAG'
    script: |
      #!/usr/bin/env bash
      cp -R client-libs/ website/client-libs
      docker build -t \
      ${_REGION}-docker.pkg.dev/${PROJECT_ID}/website/website:${_IMAGE_TAG} \
      website/. 
```

You can can find more information about the `script` field and substitutions [right here](https://cloud.google.com/build/docs/configuring-builds/run-bash-scripts#using_substitutions_with_the_script_field).

File: ops/web-build.cloudbuild.yaml [hunk 1]
```diff
@@ -17,14 +17,12 @@ steps:
...
-      - '-c'
-      - |
-        cp -R client-libs/ website/client-libs
-        docker build -t \
-        ${_REGION}-docker.pkg.dev/${PROJECT_ID}/website/website:${_IMAGE_TAG} \
-        website/. 
+    script: |
+      #!/usr/bin/env bash
+      cp -R client-libs/ website/client-libs
+      docker build -t \
```
Comment (rogerthatdev) [inline ops/web-build.cloudbuild.yaml:23]: Cloud Build will not read values for these variables. Include an `env` field to remedy: 
```
  - name: gcr.io/cloud-builders/docker
   env:
    - '_REGION=$_REGION'
    - 'PROJECT_ID=$PROJECT_ID'
    - '_IMAGE_TAG=$_IMAGE_TAG'
```

**Review by rogerthatdev — CHANGES_REQUESTED** [review]
Thanks for the contribution!

In my testing I found that there is an issue having to do with variable substitution in your changes. For example, when initiating the `web-new-build` Cloud Build trigger that uses config `ops/web-build.cloudbuild.yaml` it returns the following error:
```
invalid argument "-docker.pkg.dev//website/website:" for "-t, --tag" flag: invalid reference format
See 'docker build --help'.
ERROR
ERROR: build step 0 "gcr.io/cloud-builders/docker" failed: step exited with non-zero status: 125
```

`-docker.pkg.dev//website/website:` in the build config reads `${_REGION}-docker.pkg.dev/${PROJECT_ID}/website/website:${_IMAGE_TAG} ` but the `script` field does not  support values from the [`substitution` field of the build config](https://github.com/GoogleCloudPlatform/emblem/blob/4be1a2f76f7b684f427d74db0209506e3d60f5af/ops/web-build.cloudbuild.yaml#L30C1-L32C25) 

Instead, each step that uses the `script` field with substitution values needs to also include an `env` field that will pass values to the script via environment variables. For example, for the `ops/web-build.cloudbuild.yaml`: 
```
steps:
  # Docker Build 
  # Copying the client-libs for Docker context
  - name: gcr.io/cloud-builders/docker
   env:
    - '_REGION=$_REGION'
    - 'PROJECT_ID=$PROJECT_ID'
    - '_IMAGE_TAG=$_IMAGE_TAG'
    script: |
      #!/usr/bin/env bash
      cp -R client-libs/ website/client-libs
      docker build -t \
      ${_REGION}-docker.pkg.dev/${PROJECT_ID}/website/website:${_IMAGE_TAG} \
      website/. 
```

You can can find more information about the `script` field and substitutions [right here](https://cloud.google.com/build/docs/configuring-builds/run-bash-scripts#using_substitutions_with_the_script_field).

**Comment by smeet07** [comment]
Oh I did not know script tag does not support substitution variables. Sorry for not doing the research beforehand, I’ll make the appropriate changes

**Comment by smeet07** [comment]
@rogerthatdev need /gcbrun from you to test the api-unit-tests

**Comment by pattishin** [comment]
@smeet07 `failed unmarshalling build config ops/unit-tests.cloudbuild.yaml: json: cannot unmarshal string into Go value of type []json.RawMessage` 

Screenshot 2023-06-26 at 10 30 10 AM

To validate that the cloud build trigger above is working on your end, do setup this project on a personal GCP project (https://github.com/GoogleCloudPlatform/emblem/tree/main/docs). It will take 15-20 minutes to fully set this up via terraform and execute the trigger by simply updating anything within `content-api/` dir.

**Comment by smeet07** [comment]
> @smeet07 Screenshot 2023-06-26 at 10 30 10 AM
> 
> To validate that the `api-unit-tests` cloud build trigger is working on your end, do setup this project on a personal GCP project (https://github.com/GoogleCloudPlatform/emblem/tree/main/docs). It will take 15-20 minutes to fully set this up via terraform and execute the trigger by simply updating anything within `content-api/` dir.

I do not have a billing enabled gcp account at the moment, I'll see if I have credits left on my uni account. Meanwhile could you trigger the api-unit-tests build?

**Comment by smeet07** [comment]
There is no console output of the error, the problem might be due to #848

**Comment by smeet07** [comment]
> /gcbrun

@pattishin could you give me console output of the error

**Review by pattishin — CHANGES_REQUESTED** [review]
Hi @smeet07, running your fork locally is producing the following error. Adding some suggestions here to resolve and once those changes are in this should be ready to merge.

Screenshot 2023-07-06 at 4 07 01 PM

File: ops/web-build.cloudbuild.yaml [hunk 2]
```diff
@@ -17,14 +17,16 @@ steps:
...
+      - '_IMAGE_TAG=$_IMAGE_TAG'
+    script: |
+      #!/usr/bin/env bash
+      cp -R client-libs/ website/client-libs
+      docker build -t \
+      ${_REGION}-docker.pkg.dev/${PROJECT_ID}/website/website:${_IMAGE_TAG} \
+      website/. 
 
 # Default to us-central1 
 substitutions:
```
[the remark this answers is not in the document]
Comment (smeet07) [reply ops/web-build.cloudbuild.yaml:32]: So in the images section should I hardcode the values. Replace ${_REGION} with the hardcoded value 'us-central1', and ${_IMAGE_TAG} with '${SHORT_SHA}' ?

**Comment by smeet07** [comment]
> @smeet07 Thanks for the changes! Do you mind removing that last `substitution` block? I'll retest for you and get these changes in.

@pattishin I've made the changes

**Comment by smeet07** [comment]
@pattishin I can see that the build is still failing , what is the error?

**Review by pattishin — APPROVED** [review]
@smeet07 It's a known issue that is not related to your PR (api unit test trigger). I just verified your PR manually. Thanks for your contribution!

**Pull request merged**